### PR TITLE
[CAS-771] State flow chat domain

### PR DESCRIPTION
**Pull request by ogkuzmin**
### 🎯 Goal

Create a new ChatDomain based on `StateFlow`.

### 🛠 Implementation details

Basically it's huge PR is just wrapping new copy paste version of ChatDomain, ChatDomainImpl based on `StateFlow` in order to provide old `LiveData` based API. Main changes

1) Add new `ChatDomain` interface with builder and `ChatDomainImpl` in the offline package
2) Make old `ChatDomainImpl` as wrapper with an instance of the `ChatDomain` interface from offline package
3) Move `EventHandler` and `EventBatchUpdate` to offline and use `ChatDomainImpl` from the offline package there
4) Copy `UseCases`, `UseCaseHelper` to the offline package. Old classes are now wrappers
5) Rewrite tests to use `StateFlow` classes (some bonus fix assertions in some tests)

### ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (CMS, cookbooks, tutorial)
- [x] Reviewers added

### CODE REVIEW

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/usecase/CancelMessage.kt
Comment (ogkuzmin) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/usecase/CancelMessage.kt:10]: Use case are copy pasted from the live data package

File: stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/BaseConnectedIntegrationTest.kt
Comment (ogkuzmin) [inline stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/BaseConnectedIntegrationTest.kt:62]: Need to launch new coroutine otherwise collecting will block since `StateFlow` is never finishing flow objects

File: stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/BaseDomainTest.kt
Comment (ogkuzmin) [inline stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/BaseDomainTest.kt:238]: Same here launching new coroutine

File: stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/controller/ChannelControllerImplEventTest.kt
Comment (ogkuzmin) [inline stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/controller/ChannelControllerImplEventTest.kt:102]: It's not logically right according to our current logic to check `updatedAt`. Current logic checks `createdAt` property

File: stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/controller/ChannelControllerImplEventTest.kt
Comment (ogkuzmin) [inline stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/controller/ChannelControllerImplEventTest.kt:117]: It wasn't assert even previously :-)

File: stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/usecase/GetTotalUnreadCountImplTest.kt
Comment (ogkuzmin) [inline stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/usecase/GetTotalUnreadCountImplTest.kt:13]: Don't see any reason to keep ignored tests for deprecated use case

File: stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/usecase/SendMessageWithAttachmentsImplTest.kt
Comment (ogkuzmin) [inline stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/usecase/SendMessageWithAttachmentsImplTest.kt:41]: Again broken ignored tests for deprecated use case

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/ChatDomain.kt
Comment (leandroBorgesFerreira) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/ChatDomain.kt:590]: Shouldn't remove the `user` from this builder?

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/usecase/LoadMessageByIdImpl.kt
Comment (leandroBorgesFerreira) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/usecase/LoadMessageByIdImpl.kt:28]: Why we need to duplicate `LoadMessageByIdImpl`

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/channel/ChannelController.kt
Comment (leandroBorgesFerreira) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/channel/ChannelController.kt:105]: But why not just keep the old constructor? We still can't use any different implementation different `ChatDomainImpl`... This change is only dropping the compiler check for us.

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/usecase/CancelMessage.kt
Comment (leandroBorgesFerreira) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/usecase/CancelMessage.kt:10]: But why we need those copy and paste, instead of a move?

**Comment by leandroBorgesFerreira** [comment]
I didn't understand why we need to duplicate the usecases and UsecaseHelper. Wouldn't it be confusing to have 2 classes for the same job? I don't see diferences between one and the other.

**Comment by ogkuzmin** [comment]
@leandroBorgesFerreira When we will have a standalone artifact for the offline package if we move use cases only there then it will break customers’ codebases, they will be forced to fix imports and include this new artifact into their dependencies. So that’s why basically all public classes and funcs in the livedata package must be kept and we have to provide also versions in the offline package

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/ChatDomain.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/ChatDomain.kt:590]: We have to keep old behaviour in new version. Even though this user has been deprecated it has an impact on logic inside `ChatDomainImpl`. This PR shouldn't change any logic, but provide all old without `LiveData` dependency for `ChatDomain`.

File: UNRELEASED_CHANGELOG.md
Comment (zsmb13) [inline UNRELEASED_CHANGELOG.md:46]: I'm not sure where and when we want to document these new APIs, but we should be linking to them here in the changelog.

**Review by zsmb13 — COMMENTED** [review]
Huge amount of work, really well done overall 👍. Got some things I'd clean up / didn't quite understand, see comments.

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/usecase/LoadMessageByIdImpl.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/usecase/LoadMessageByIdImpl.kt:28]: The same reason. It's interface from the `livedata` package is public and we must keep it

File: stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/ChatDomainImplCreateChannelTest.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/ChatDomainImplCreateChannelTest.kt:24]: I will do it in next PR

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/usecase/GetChannelControllerImpl.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/usecase/GetChannelControllerImpl.kt:25]: We need copying even those use cases that return simple `Call` because they have public interfaces from the `livedata` package and they must be kept, while versions from the `offline` package shouldn't have any dependency from `livedata`

Comment (zsmb13) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/usecase/GetChannelControllerImpl.kt:25]: Got it, thank you! So we need these for now, and then if we remove the `livedata.ChatDomain.useCases` from public API, we can kill all the use cases in there, and just merge the mapping code into `livedata.ChatDomain` instead. That'll clean up a bunch of small files.

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/usecase/SendMessageImpl.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/usecase/SendMessageImpl.kt:35]: Merge issue. Fixed that and also removed this use case from offline version

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/usecase/UseCaseHelper.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/usecase/UseCaseHelper.kt:10]: Made it deprecated

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/querychannels/QueryChannelsController.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/querychannels/QueryChannelsController.kt]: Good catch. Expose all methods and properties as it was before

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/querychannels/QueryChannelsController.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/querychannels/QueryChannelsController.kt]: Good catch. Expose all methods and properties as it was before

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/thread/ThreadController.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/thread/ThreadController.kt]: Made it internal

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/usecase/CancelMessage.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/usecase/CancelMessage.kt]: Yes, you're right. Remove this redundant interfaces and keep only classes

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/usecase/CancelMessage.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/usecase/CancelMessage.kt:19]: Why not, if previous interface was public and we must keep them?

Comment (zsmb13) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/usecase/CancelMessage.kt:19]: You're right, we need these to keep the existing ones in the package they are, my bad 👍

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/usecase/UseCaseHelper.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/usecase/UseCaseHelper.kt]: Yes, I agree. Removed it and hid use cases from the offline package

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/channel/ChannelController.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/channel/ChannelController.kt]: fixed

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomain.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomain.kt]: But what do you suggest? It's still present in public interface? I don't want to keep any logic in livedata version of `ChatDomain`. We'll remove it from interfaces with deprecation approach later

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomain.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomain.kt:656]: Yes, it was intended, we'll expose it by default later. So customers who don't need any live-data dependency won't use them

File: UNRELEASED_CHANGELOG.md
Comment (ogkuzmin) [reply UNRELEASED_CHANGELOG.md:46]: So you prefer not to mention this offline classes right now?

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/channel/ChannelController.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/channel/ChannelController.kt]: Done

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomain.kt
Comment (zsmb13) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomain.kt]: We could mark these as `@InternalStreamChatApi` if we still want the `livedata` interface to call into it, but not anyone to start using the new API here

File: UNRELEASED_CHANGELOG.md
Comment (zsmb13) [reply UNRELEASED_CHANGELOG.md:48]: I think when we ship these as public API & notify clients about it in the changelog, we should have all the relevant docs linked. Doesn't necessarily have to be added to the changelog on this same PR, I'm just generally asking about what the plan to document these APIs is.

Comment (ogkuzmin) [reply UNRELEASED_CHANGELOG.md:48]: I created a new Jira task in epic. So we can postpone this discussion

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomain.kt
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomain.kt]: I got your point. Removed the deprecated annotation and put `@InternalStreamChatApi`

**Review by zsmb13 — APPROVED** [review]
LGTM now, thanks for addressing all my comments. Again, great work! Please wait for a second review on this one before merging, as there are a lot of changes here.

**Pull request merged**
